### ISSUE TITLE
banner drops last column of pixels

### ISSUE THREAD

**Issue by JohnCC330** [robhagemans/monobit#13]
It might just be my bad luck with the first font I selected, but I noticed that monobit-banner is dropping the last 'column' of bits:

```
................
................
................
................
................
................
@@.@@@...@@.@@@.
.@@...@...@@...@
.@....@...@....@
.@....@...@....@
.@....@...@....@
.@....@...@....@
.@....@...@....@
@@@..@@@.@@@..@@
................
................
................

```
This uses `apple-mac/Espy_Serif_14.yaff` (from[ The hoard](http://robhagemans.github.io/monobit/)), and text `n n`. This is solvable by adding a space to the input text `n n `:

```
..................
..................
..................
..................
..................
..................
@@.@@@...@@.@@@...
.@@...@...@@...@..
.@....@...@....@..
.@....@...@....@..
.@....@...@....@..
.@....@...@....@..
.@....@...@....@..
@@@..@@@.@@@..@@@.
..................
..................
..................

```
I'm not sure this is actually the result of an issue with the font, or in the `monobit-banner` code.
It's not world-shattering...

**Comment by robhagemans** [comment]
Hi, thanks for reporting!

I'll have a look. That output looks weird, so I'd say the issue is with the
renderer - even if the font were badly designed (and I'm sure the Apple
fonts are not!) that should not happen...

**Comment by robhagemans** [comment]
I just checked and actually it's as expected - it looks a bit weird but there's a reason for that. Not as weird as I first thought by the way, I think my email client showed the output not monospaced which deformed it further.

What's happening is that the `n` in Etsy Serif has a negative right sidebearing (`right-bearing: -1` in the glyph definition in the yaff file), which means it shifts left by one column after drawing the last `n`. Effectively the last pixel of the serif is drawn in the margin, but by default there is no margin, so it drops off.

Compare:

    $ ./banner.py 'n n' --font apple-mac/Espy_Serif_14.yaff --margin 2x0
    ....................
    ....................
    ....................
    ....................
    ....................
    ....................
    ..@@.@@@...@@.@@@...
    ...@@...@...@@...@..
    ...@....@...@....@..
    ...@....@...@....@..
    ...@....@...@....@..
    ...@....@...@....@..
    ...@....@...@....@..
    ..@@@..@@@.@@@..@@@.
    ....................
    ....................
    ....................

In this rendering, the leftmost and rightmost two columns are margin. Some glyphs will also draw into the left margin. 

I do think the bearings on this font look very tight as it stands. I wonder how it was actually rendered on the mac - maybe I'm misinterpreting the metrics in the original `NFNT`. Any screenshots welcome :)

**Comment by JohnCC330** [comment]
> What's happening is that the n in Etsy Serif has a negative right sidebearing (right-bearing: -1 in the glyph definition in the yaff file), which means it shifts left by one column after drawing the last n. Effectively the last pixel of the serif is drawn in the margin, but by default there is no margin, so it drops off.

First apologies for the delay... Again on the job.

I suspected the `right-bearing` when viewing the font sources... Actually quite a few fonts drop columns at the right, and it might be helpful to add a warning to the docs, that even with no special 'effects', some fonts might render incompletely, indicating the need for added margins. 

In fact, it might be advisable, in the interest of user-friendliness, to extend the margins automatically?

Again, just suspicion, I think the apple fonts need an extra 'implicit' column between characters (not defined in the font). If not, combinations like 'lll' (3 L's) produce a solid 'black' block, and things like 'rm' seem to blend or even overwrite.

Ah well, fonts are not easy to begin with :)

Groetjes
John

**Comment by robhagemans** [comment]
I had a closer look and think I know what's going on. _Espy Sans 9pt_ has a value of `-1` for `kernMax` in the `NFNT` header:

> Maximum kerning. An integer value that specifies the distance from the font rectangle's glyph origin to the left edge of the font rectangle, in pixels. If a glyph in the font kerns to the left, the amount is represented as a negative number. If the glyph origin lies on the left edge of the font rectangle, the value of the kernMax field is 0. This value is represented by the kernMax field in the FontRec data type.

It looks like there's a bug/misinterpretation in how I apply that in the converter in that it correctly applies this as an offset to the left side bearing, but then doesn't adjust the right side bearing accordingly. So everything comes out one pixel too tight. This also affects the other _Espy_ s and _Palatino_, _Times_, _Symbol_, and _Helvetica_ (but not Susan Kare's fonts which all have zero `kernMax`). They seem to render much better if I make the adjustment. I'll fix the converter code and re-extract them.

Re your suggestion on extending the margins, I see where you're coming from. My worry is that doing too much stuff automatically is going to mask actual font conversion problems (like this one). I kind of want to  give/have pixel-level control in this utility. Maybe the solution is to have the default setting add a margin as needed, but with a different default background, such as spaces. I don't know yet. Similarly, I'm thinking of having some options to tighten or loosen bearings, or to completely override them. But the defaults should just be what the font actually defines... 

Groetjes & bedankt
Rob

**Comment by robhagemans** [comment]
For reference, this is how they _should_ look:  

<img src='https://lowendmac.com/wp-content/uploads/espy.png'>

**Comment by JohnCC330** [comment]
> It looks like there's a bug/misinterpretation in how I apply that in the converter in that it correctly applies this as an offset to the left side bearing, but then doesn't adjust the right side bearing accordingly. So everything comes out one pixel too tight. This also affects the other Espy s and Palatino, Times, Symbol, and Helvetica (but not Susan Kare's fonts which all have zero kernMax). They seem to render much better if I make the adjustment. I'll fix the converter code and re-extract them.

Great! I really liked the fonts at first sight, but the strange spacing issues had me  wondering. Some tests I did also show that 'algorthmic' bolding of a font causes letter spacing to be consumed. I suspect that italicizing might cause similar headaches.

> Re your suggestion on extending the margins, I see where you're coming from. My worry is that doing too much stuff automatically is going to mask actual font conversion problems (like this one). I kind of want to give/have pixel-level control in this utility. Maybe the solution is to have the default setting add a margin as needed, but with a different default background, such as spaces. I don't know yet. Similarly, I'm thinking of having some options to tighten or loosen bearings, or to completely override them. But the defaults should just be what the font actually defines...

I understand with the potential issues - but maybe the 'user friendliness' should win out here? Or maybe more useful in both cases: As you propose, a default of auto-correction, but print a warning if 'automatic'  and 'strict' don't produce the same result?

Anyway, again, I appreciate the work you're doing. Really - 30+ years ago I designed and built an electronic terminal with its own (bit-mapped of course, and mono-spaced) font (for ham-radio communications).

I look forward to the updated version. BTW, if I can be of help...

John

**Comment by robhagemans** [comment]
Fixed in commit aec4782 - I also updated the repo.

**Comment by robhagemans** [comment]
This was fixed with v0.38
